vnet/net.go

```diff
@@ -0,0 +1,645 @@
+	}
+
+	space := end + 1 - start
+	offset := rand.Intn(space)
+	for i := 0; i < space; i++ {
+		port := ((offset + i) % space) + start
+
+		err := v.allocateLocalAddr(ip, port)
+		if err == nil {
+			return port, nil
+		}
+	}
+
+	return -1, fmt.Errorf("port space exhausted")
+}
+
+// NetConfig ...
+type NetConfig struct {
+	// StaticIP is a static IP address to be assigned for this Net. If nil,
+	// the router will automatically assign an IP address.
+	StaticIP string
+}
+
+// Net represents a local network stack euivalent to a set of layers from NIC
+// up to the transport (UDP / TCP) layer.
+type Net struct {
+	v   *vNet
+	ifs []*Interface
+}
+
+// NewNet creates an instance of Net.
+// If config is nil, the virtual network is disabled. (uses corresponding
+// net.Xxxx() operations.
+// By design, it always have lo0 and eth0 interfaces.
+// The lo0 has the address 127.0.0.1 assigned by default.
+// IP address for eth0 will be assigned when this Net is added to a router.
+func NewNet(config *NetConfig) *Net {
+	if config == nil {
+		ifs := []*Interface{}
+		if orgIfs, err := net.Interfaces(); err == nil {
+			for _, orgIfc := range orgIfs {
+				ifc := NewInterface(orgIfc)
+				if addrs, err := orgIfc.Addrs(); err == nil {
+					for _, addr := range addrs {
+						ifc.AddAddr(addr)
+					}
+				}
+
+				ifs = append(ifs, ifc)
+			}
+		}
+
+		return &Net{ifs: ifs}
+	}
+
+	lo0 := NewInterface(net.Interface{
+		Index:        1,
+		MTU:          16384,
+		Name:         "lo0",
+		HardwareAddr: nil,
+		Flags:        net.FlagUp | net.FlagLoopback | net.FlagMulticast,
+	})
+	lo0.AddAddr(&net.IPNet{
+		IP:   net.ParseIP("127.0.0.1"),
+		Mask: net.CIDRMask(8, 32),
+	})
+
+	eth0 := NewInterface(net.Interface{
+		Index:        2,
+		MTU:          1500,
+		Name:         "eth0",
+		HardwareAddr: newMACAddress(),
+		Flags:        net.FlagUp | net.FlagMulticast,
+	})
+
+	v := &vNet{
+		interfaces: []*Interface{lo0, eth0},
+		staticIP:   net.ParseIP(config.StaticIP),
+		udpConns:   map[string]*UDPConn{},
+	}
+
+	return &Net{
+		v: v,
+	}
+}
+
+// Interfaces returns a list of the system's network interfaces.
+func (n *Net) Interfaces() ([]*Interface, error) {
+	if n.v == nil {
+		return n.ifs, nil
+	}
+
+	return n.v.getInterfaces()
+}
+
+// InterfaceByName returns the interface specified by name.
+func (n *Net) InterfaceByName(name string) (*Interface, error) {
+	if n.v == nil {
+		for _, ifc := range n.ifs {
+			if ifc.Name == name {
+				return ifc, nil
+			}
+		}
+
+		return nil, fmt.Errorf("interface %s not found", name)
+	}
+
+	return n.v.getInterface(name)
+}
+
+// ListenPacket announces on the local network address.
+func (n *Net) ListenPacket(network string, address string) (net.PacketConn, error) {
+	if n.v == nil {
+		return net.ListenPacket(network, address)
+	}
+
+	return n.v.listenPacket(network, address)
+}
+
+// ListenUDP acts like ListenPacket for UDP networks.
+func (n *Net) ListenUDP(network string, locAddr *net.UDPAddr) (UDPPacketConn, error) {
```

Comment:
this should actually return a pointer to the conn to match the net interface: https://golang.org/pkg/net/#ListenUDP

Reply:
Yeah, that's the thing I have debated.
In order for vnet.ListenUDP() to support both `native` (returns *net.UDPConn) and `virtual` (returns *vnet.UDPConn), we need to introduce an interface that allows both of those pointers. For that reason, vnet.UDPPacketConn had to be introduced.

In many cases, we could get away with using ListenPacket() method and I even thought not supporting vnet.ListenUDP(). (Google encourage users to use ListenPacket() also. But I guess, vnet.ListenUDP() make it a bit easier to migrate to vnet.

Reply:
I updated README at @ea6c83d. Does it make sense to you?

Reply:
it does makes sense! and I can see why ListenPacket will be preferred too, we can probably change that later.